python/cudf/cudf/dataframe/categorical.py

```diff
@@ -92,20 +113,26 @@ def __init__(self, **kwargs):
         ordered : bool
             whether the categorical has a logical ordering (e.g. less than)
         """
-        categories = kwargs.pop("categories")
-        ordered = kwargs.pop("ordered")
-        dtype = pd.core.dtypes.dtypes.CategoricalDtype(categories, ordered)
-        kwargs.update({"dtype": dtype})
+        from cudf.dataframe.string import StringColumn
+
+        ordered = bool(kwargs.pop('ordered'))
+        categories = kwargs.pop('categories', [])
+        # Default to String dtype if len(categories) == 0, like pandas does
+        categories = columnops.as_column(categories) if len(categories) > 0 \
+            else StringColumn(data=nvstrings.to_device([]), null_count=0)
+
+        dtype = CategoricalDtype(ordered=ordered)
+        kwargs.update({'dtype': dtype})
         super(CategoricalColumn, self).__init__(**kwargs)
-        self._categories = tuple(categories)
-        self._ordered = bool(ordered)
+        self._categories = categories
+        self._ordered = ordered
 
     def serialize(self, serialize):
         header, frames = super(CategoricalColumn, self).serialize(serialize)
-        assert "dtype" not in header
-        header["dtype"] = serialize(self._dtype)
-        header["categories"] = self._categories
-        header["ordered"] = self._ordered
+        assert 'dtype' not in header
+        header['dtype'] = serialize(self._dtype)
+        header['categories'] = self._categories.to_array()
```

Comment:
You actually shouldn't do this. We have CUDA serialization in place for handling the various Column objects and it can fall back to CPU based serialization when CUDA serialization isn't available.

@mrocklin @pentschev Does the below look correct?
```suggestion
        header["category_buffer"], category_frames = serialize(self._categories)
        header["category_frame_count"] = len(category_frames)
        frames.extend(category_frames)
```

Reply:
Need to update the deserialize method accordingly as well.

Reply:
Thanks for the suggestion on serialize. Do you have an example of what deserialize should look like too, or a pointer where I can see a description of the serialization format?

Another thing that's unclear is how/whether dictionaries are deduplicated when serialized. Arrow writes a dictionary id in the schema metadata, but it seems here we'd serialize the same dictionary multiple times.

Reply:
Deserialize should look something like this: https://github.com/rapidsai/cudf/blob/branch-0.9/python/cudf/cudf/dataframe/numerical.py#L57

Reply:
Sorry for the late reply. It looks a little strange that now `self._categories` is being stored twice, once as an array and once as a serialized buffer, is that on purpose?

Reply:
It shouldn't only be stored in `header["category_buffer"]` unless I'm missing something, `header['categories']` is being deleted above.

Reply:
Ah damn, I've mistaken your suggested changes as a continuation of the changes in this PR. Sorry about that.

The suggested changes do look correct, yes. And well observed about the missing deserialize update.

Reply:
I double-checked the dask.distributed docs and did some debugging to verify, does this look OK?

https://github.com/rapidsai/cudf/blob/5a08faf9cfffd38f3403a2f9261df7a4d0a0bce6/python/cudf/cudf/dataframe/categorical.py#L131-L166

I wasn't sure if I should `assert 'dtype' not in header['categories']`, or if it would be added by the call to `serialize()`, so I went with the non-throwing option.

Reply:
Why do we need to check and add the dtype to the categories header? Serializing the column represented by `self._categories` should handle that, no?

Reply:
I was trying to follow the example of serializing the column above, where it does this after calling super.serialize():
```python
assert 'dtype' not in header
```

When I was debugging, it looked like this call _does not_ add a dtype to `header['categories']`, but I wasn't sure if it never does vs. sometimes does:
```python
header['categories'], category_frames = serialize(self._categories) 
```

Reply:
I believe it puts the dtype in only if it needs it, i.e. NumericalColumn and DatetimeColumn that can have different dtypes within the same class. Whereas StringColumn it doesn't need to.